## Introduction

This project is meant to help gain better understanding as to what types of ads draw the most ad revenue. My goal in this project is to demonstrate the skills I learned in this module of DataQuest.

1. Here we open the data sets and save them as lists of lists.

In [1]:
from csv import reader

### Google Play Dataset###
opened = open('googleplaystore.csv')
read = reader(opened)
android = list(read)

android_header = android[0]
android = android[1:]


### IOS Dataset ###
opened = open('AppleStore.csv')
read = reader(opened)
ios = list(read)

ios_header = ios[0]
ios = ios[1:]

Explore both data sets using ```explore_data()``` function

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
print(android_header)
print('\n')
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [4]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


Since both datasets contain header rows, we see that the android dataset contains 10841 observations and 13 columns while the ios dataset contains 7197 observations and 16 columns.

In [5]:
explore_data(android,0,1)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']




The [google play dataset](https://www.kaggle.com/lava18/google-play-store-apps) has useful variables to analyze such as App, and Genres.




In [6]:
explore_data(ios,0,1)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']




The most useful variables in the [ios dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) would be track_name, currency, price, rating_count_tot, rating_count_ver, user_rating, and prime_genre.

## Deleting Wrong Data

In [7]:
print(android[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [8]:
del(android[10472]) # only run this one time

## Checking the data has duplicates

The google dataset we possess many duplicates of the same app. To show some below:

In [9]:
for app in android:
    name = app[0]
    if name == 'Facebook':
        print(app)

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']


In [10]:
for app in android:
    name = app[0]
    if name == 'YouTube':
        print(app)

['YouTube', 'VIDEO_PLAYERS', '4.3', '25655305', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Video Players & Editors', 'August 2, 2018', 'Varies with device', 'Varies with device']
['YouTube', 'VIDEO_PLAYERS', '4.3', '25623548', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Video Players & Editors', 'August 2, 2018', 'Varies with device', 'Varies with device']


In [11]:
### Count number of duplicates

duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps))

Number of duplicate apps: 1181


To determine how we will remove duplicates, let's take a look at the review counts for each app. This data can give us an idea of which verision is the most recent.

In [12]:
for app in android:
    name = app[0]
    if name == 'ESPN':
        print(app)

['ESPN', 'SPORTS', '4.2', '521138', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone 10+', 'Sports', 'July 19, 2018', 'Varies with device', '5.0 and up']
['ESPN', 'SPORTS', '4.2', '521138', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone 10+', 'Sports', 'July 19, 2018', 'Varies with device', '5.0 and up']
['ESPN', 'SPORTS', '4.2', '521138', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone 10+', 'Sports', 'July 19, 2018', 'Varies with device', '5.0 and up']
['ESPN', 'SPORTS', '4.2', '521140', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone 10+', 'Sports', 'July 19, 2018', 'Varies with device', '5.0 and up']
['ESPN', 'SPORTS', '4.2', '521140', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone 10+', 'Sports', 'July 19, 2018', 'Varies with device', '5.0 and up']
['ESPN', 'SPORTS', '4.2', '521140', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone 10+', 'Sports', 'July 19, 2018', 'Varies with device', '5.0 and up']
['ESPN', '

 We can see that the differing factor between each row is the 4th column, which represents the number of reviews.
 Deleting the duplicates except for the row with the highest number should therefore leave us with the most recent version of each app.

We will do this by:
    - creating a dictionary containing the highest review count for each app
    - using that dictionary to make a new list of apps only containing the occurence with the highest reviews

## Removing Duplicate Entries

1. Initialize two lists: one for the new clean dataset and one for remembering the apps we added
2. Loop through the android dataset and add the row, or app, to the clean dataset if:
    * The number of reviews matches that of the number in the reviews_max dictionary
    * The app has not already been added

In [13]:
### Create dictionary with key: unique app name and value: highest review count

# Create empty dictionary
reviews_max = {}


for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))


9659


By exploring the cleaned dataset, we can now comfirm that there are 9659 rows.

In [14]:
# Use created dictionary to remove duplicate rows

android_clean = []  # Store new cleaned dataset
already_added = []  # Store app names

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)

explore_data(android_clean,0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


## Removing Non-English Apps

In [15]:
# Function to check if word is English
# Returns True if word is English
# Returns False if not

def english_checker(name):
    for letter in name:
        if ord(letter) > 127:
            return False
    return True

We see that our function works mostly as inteded. However, there are still cases that need fixing. The last two examples below represent possible English titles that contain characters that have an ASCII over 127, thus get classified as non English. 

In [16]:
# Test if function works

print(english_checker('Instagram'))
print(english_checker('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_checker('Docs To Go™ Free Office Suite'))
print(english_checker('Instachat 😜'))

True
False
False
False


In [17]:
# Changing the function we made earlier

def english_checker(name):
    
    n = 0
    
    for letter in name:
        if ord(letter) > 127:
            n+=1
                    
    if n > 3:
        return False
    else:
        return True

In [18]:
print(english_checker('Instagram'))
print(english_checker('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_checker('Docs To Go™ Free Office Suite'))
print(english_checker('Instachat 😜'))

True
False
True
True


With our improved function, we use it to filter both android and ios datasets. After filtering for non-English names, we are left with 9614 and 6183 rows respectively.

In [19]:
android_english = []
ios_english = []

# filter out non-English android apps

for app in android_clean:
    name = app[0]
    if english_checker(name):
        android_english.append(app)

# filter out non-English ios apps

for app in ios:
    name = app[1]
    if english_checker(name):
        ios_english.append(app)
        
        
explore_data(android_english,0,3, True)
print('\n')
explore_data(ios_english,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

## Isolating the Free Apps

In [20]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)

for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
print(len(android_final))
print(len(ios_final))


8864
3222


## Most Common Apps by Genre

### Part One

Our goal is to determine the kind of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps. 

To minimize risk and overhead, our validation strategy for our app idea is:
    1. Build a minimal Android version of the app, and add it to Google Play.
    2. If the app has a good response from users, we develop it further.
    3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store

In [21]:
print(android_header,'\n')
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


We want to generate frequency tables to find out what are the most common genres in each market. Useful columns could be Genres and prime_genre.

### Part Two

In [22]:
#build function to generate frequency tables that show percentages

def freq_table(dataset, index):
    frequency_table = {}
    total = 0  # tallying total number for calculating percentages
    
    for row in dataset:
        total += 1
        datapoint = row[index]
        if datapoint in frequency_table:
            frequency_table[datapoint] += 1
        else:
            frequency_table[datapoint] = 1
            
    percentages = {}
    for key in frequency_table:
        percentages[key] = (frequency_table[key]/total)*100
        
    return percentages
        
        
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### Part Three

Let's examine the `prime_genre` column from the ios dataset

In [23]:
display_table(ios_final, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


The most common genre by far is 'Games', which is about 58.16% of the dataset. The next genre is 'Entertainment', which makes up about 7.88 
One observation we can draw from this would be that most apps on the app store are designed for entertainment and fun. There are significantly less apps meant for practical purposes (education, shopping, utilities, productivity, lifestyle).
It is hard to recommend a genre off of this frequency table alone. A large number of apps in a genre does not neccesarily indicate a large number of users for that genre.


In [24]:
display_table(android_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

Looking at `Category` column of the Android dataset, we see that the most common category is Family, which consists of 18.908% of the data. The google play dataset seems to consist of apps more for practical purposes rather than for entertainment.

In [25]:
display_table(android_final, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

In the `Genre` category, the Tools genre is the most common, at 8.45%. Again, we notice that there are more practical apps compared to the App Store.
These frequency tables told us what app genres are most frequent but not which ones have the most users.

## Most Popular Apps by Genre on the App Store

In [26]:
prime_genre = freq_table(ios_final,-5)
for genre in prime_genre:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:
            num_ratings = float(app[5])
            total += num_ratings
            len_genre += 1
            
    avg_rting = total/len_genre
    print(genre, avg_rting)

Social Networking 71548.34905660378
Photo & Video 28441.54375
Games 22788.6696905016
Music 57326.530303030304
Reference 74942.11111111111
Health & Fitness 23298.015384615384
Weather 52279.892857142855
Utilities 18684.456790123455
Travel 28243.8
Shopping 26919.690476190477
News 21248.023255813954
Navigation 86090.33333333333
Lifestyle 16485.764705882353
Entertainment 14029.830708661417
Food & Drink 33333.92307692308
Sports 23008.898550724636
Book 39758.5
Finance 31467.944444444445
Education 7003.983050847458
Productivity 21028.410714285714
Business 7491.117647058823
Catalogs 4004.0
Medical 612.0


Looking at the data, the most popular genres seem to be Social Networking, Reference, Weather, Navigation, and Education. However, these genres seem to be dominated by giant players which make the markets hard to penetrate. For example, social networking is dominated by giants such as Facebook, Twitter etc. Navigation mainly consists of Waze and Google Maps. Popular genres which do not seem to have such monopolized market shares would be Games, Health & Fitness, and Productivity.

## Most Popular Apps by Genre on Google Play

In [28]:
display_table(android_final, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


 We have data about the number of installs for the Google Play market, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.)

In [33]:
category_table = freq_table(android_final,1)
for category in category_table:
    total = 0          # store sum of installs specific to each genre
    len_category = 0   # store number of apps specific to each genre
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            num_installs = app[5]
            repl_install = num_installs.replace('+','')
            repl_install = repl_install.replace(',','')
            float_install = float(repl_install)
            total += float_install
            len_category += 1
    
    avg_install_no = total/len_category
    print(category, avg_install_no)
    

ART_AND_DESIGN 1986335.0877192982
AUTO_AND_VEHICLES 647317.8170731707
BEAUTY 513151.88679245283
BOOKS_AND_REFERENCE 8767811.894736841
BUSINESS 1712290.1474201474
COMICS 817657.2727272727
COMMUNICATION 38456119.167247385
DATING 854028.8303030303
EDUCATION 1833495.145631068
ENTERTAINMENT 11640705.88235294
EVENTS 253542.22222222222
FINANCE 1387692.475609756
FOOD_AND_DRINK 1924897.7363636363
HEALTH_AND_FITNESS 4188821.9853479853
HOUSE_AND_HOME 1331540.5616438356
LIBRARIES_AND_DEMO 638503.734939759
LIFESTYLE 1437816.2687861272
GAME 15588015.603248259
FAMILY 3695641.8198090694
MEDICAL 120550.61980830671
SOCIAL 23253652.127118643
SHOPPING 7036877.311557789
PHOTOGRAPHY 17840110.40229885
SPORTS 3638640.1428571427
TRAVEL_AND_LOCAL 13984077.710144928
TOOLS 10801391.298666667
PERSONALIZATION 5201482.6122448975
PRODUCTIVITY 16787331.344927534
PARENTING 542603.6206896552
WEATHER 5074486.197183099
VIDEO_PLAYERS 24727872.452830188
NEWS_AND_MAGAZINES 9549178.467741935
MAPS_AND_NAVIGATION 4056941.774193